Host_assistance calibration support

doronhi opened the pull request:
Added support for host-assisted on-chip-calibration and tare-calibration.
The flaw is to call the same functions, run_on_chip_calibration() and run_tare_calibration(), as before.
An empty result vector implies the need to provide the calibration process with a depth frame using the newly added function: add_calibration_frame().
The process is done when  add_calibration_frame() returns a calibration vector.

Workflow is demonstrated in wrappers\python\examples\depth_auto_calibration_example.py

daversintel committed:
First version.

daversintel committed:
1. Fixed step count error for OCC FL.
2. Added time out for starting.

daversintel committed:
More on previous check-in.

daversintel committed:
Not use extra menu item but a check box : host assitance for version 3.

daversintel committed:
1. Use pre selected resolution, which was in previous check-in laready.
2. Use centrl 1/5 to calculate the results.

daversintel committed:
progress bar.

daversintel committed:
FIxed starting point for OCC.

daversintel committed:
More bug fix for previous check-in.

daversintel committed:
Fixed counter init bug.

daversintel committed:
Fixed data size error.

daversintel committed:
1. Added timeout for fetching frames.
2. No more 256 frames waiting.
3. Update progress.

daversintel committed:
More progress bar.

daversintel committed:
Fix fetcing frame problem.

daversintel committed:
Merge remote-tracking branch 'origin/development' into Calib3

daversintel committed:
Fixed json file and progress bar problems.

daversintel committed:
More log for tare and total frame for tare.

daversintel committed:
No preset before to get calib result and table.

daversintel committed:
Merge remote-tracking branch 'origin/development' into Calib3.

daversintel committed:
Merge remote-tracking branch 'origin/development' into Calib3

# Conflicts:
#	common/on-chip-calib.cpp
#	src/ds5/ds5-auto-calibration.cpp

daversintel committed:
Merge remote-tracking branch 'origin/development' into Calib3

# Conflicts:
#	common/on-chip-calib.cpp

daversintel committed:
Fixed preet change problem.

daversintel committed:
Fixed some warning.

daversintel committed:
Preset for all tares

daversintel committed:
Fixed tare calibration crash problem caused by preset change.

daversintel committed:
Replaced perk verification to rectangle verification.

daversintel committed:
Using tare calibration status.

daversintel committed:
Added tare health check.

daversintel committed:
Merge remote-tracking branch 'origin/development' into Calib3

daversintel committed:
Enabled log for debug purpose.

daversintel committed:
Undid the debug log.

daversintel committed:
1. Use 20 frames to feedback depth in tare interactive mode.
2. Added heak check for tare.

daversintel committed:
More logs.

daversintel committed:
1. Removed OCC-FL in RS View.
2. Renamed OCC-ALL to OCC Extended and OCC Py/Rx to OCC.
3. Set the default to OCC.

daversintel committed:
API doc update.

daversintel committed:
Merge remote-tracking branch 'origin/development' into Calib3

daversintel committed:
Fixed another OCC All.

daversintel committed:
Merge remote-tracking branch 'origin/development' into Calib3

# Conflicts:
#	common/on-chip-calib.cpp
#	common/on-chip-calib.h

daversintel committed:
Merge remote-tracking branch 'origin/development' into Calib3

daversintel committed:
Moved to libRS 2.44.0 for target based focal length calibration from host only.

daversintel committed:
Merge remote-tracking branch 'origin/development' into Calib3

daversintel committed:
Merge remote-tracking branch 'origin/development' into Target_Based_Focal_Length_Calibration

daversintel committed:
Merge remote-tracking branch 'origin/development' into Calib3

daversintel committed:
Merge remote-tracking branch 'origin/development' into Target_Based_Focal_Length_Calibration

daversintel committed:
Merge remote-tracking branch 'origin/development' into Calib3

daversintel committed:
Merge remote-tracking branch 'origin/development' into Target_Based_Focal_Length_Calibration

daversintel committed:
Merge remote-tracking branch 'origin/development' into Calib3

daversintel committed:
initial UI, and ROI

daversintel committed:
UI Improvement.

daversintel committed:
Dummy uvmapping calib.

daversintel committed:
Merge branch 'Target_Based_Focal_Length_Calibration' into UVMapping

# Conflicts:
#	common/model-views.cpp
#	common/on-chip-calib.cpp
#	common/on-chip-calib.h

daversintel committed:
UVMapping partial implementation.

daversintel committed:
More on previous one.

daversintel committed:
More emitter enabled check before to set/get it.

daversintel committed:
Merge remote-tracking branch 'origin/development' into Calib3

daversintel committed:
Merge remote-tracking branch 'origin/development' into UVMapping

daversintel committed:
Merge remote-tracking branch 'origin/development' into Target_Based_Focal_Length_Calibration

daversintel committed:
Added more logs for debugging Tare HC.

daversintel committed:
Fixed another emitter related problem.

daversintel committed:
Merge remote-tracking branch 'origin/development' into Target_Based_Focal_Length_Calibration

daversintel committed:
Merge remote-tracking branch 'origin/development' into UVMapping

daversintel committed:
Merge remote-tracking branch 'origin/development' into Calib3

daversintel committed:
Corrected the Z from HC formula.

daversintel committed:
Tare HC  update.

daversintel committed:
Merge remote-tracking branch 'origin/development' into Calib3

daversintel committed:
calib3 version starts from 2.44.33.1.

daversintel committed:
Merge remote-tracking branch 'origin/development' into Target_Based_Focal_Length_Calibration

daversintel committed:
FL calibraton version starts from 2.44.22.1.

daversintel committed:
Merge remote-tracking branch 'origin/development' into UVMapping

daversintel committed:
UVMapping calibration version startds from 2.44.44.1.

daversintel committed:
Merge branch 'Target_Based_Focal_Length_Calibration' into UVMapping

# Conflicts:
#	include/librealsense2/rs.h

daversintel committed:
Fixed a typo.

daversintel committed:
Partial check in:
   1. Added RGB8 to Y8 convertion in order to support RGB8 format.
   2. Added color image undistorsion.
   3. Added uvmapping_calib class partially.

daversintel committed:
in progress.

daversintel committed:
UVMapping Calibration.

daversintel committed:
1. Log info.
2. ROI normalization.
3. UI.

daversintel committed:
1. RSV log.
2. Use 30 frames.

daversintel committed:
Use 25 frames.

daversintel committed:
No UVMapping calibration for camera without rgb camera.

daversintel committed:
Merge branch 'UVMapping' into Calib3

# Conflicts:
#	common/on-chip-calib.cpp
#	common/on-chip-calib.h
#	include/librealsense2/h/rs_device.h
#	include/librealsense2/hpp/rs_device.hpp
#	include/librealsense2/rs.h
#	src/algo.cpp
#	src/ds5/ds5-auto-calibration.cpp
#	src/rs.cpp

daversintel committed:
Fixed merge problems.

daversintel committed:
Fixed undistorting bug: 3 channels.

daversintel committed:
Start streaming initially.

daversintel committed:
Stop viewer.

daversintel committed:
Merge remote-tracking branch 'origin/development' into Calib3

daversintel committed:
Fixed stop problem.

daversintel committed:
MInor changes.

daversintel committed:
Worked around streaming issues.

daversintel committed:
3D view or not.

daversintel committed:
Added py px only option for uvmapping calibration.

daversintel committed:
Focal Length Plus.

daversintel committed:
Update right camera by ratio.

daversintel committed:
Fixed a typo.

daversintel committed:
Don't stop streming after uvmap calibration.

daversintel committed:
Fixed a typo.

daversintel committed:
Fixed soem crashes.

daversintel committed:
Merge remote-tracking branch 'origin/development' into Calib3

daversintel committed:
Temporary solution for reloading rgb camera calibration data.

daversintel committed:
Merge remote-tracking branch 'origin/development' into Calib3

daversintel committed:
Minor  var name updates.

daversintel committed:
Typo.

daversintel committed:
Merge remote-tracking branch 'origin/development' into Calib3

# Conflicts:
#	include/librealsense2/rs.h

daversintel committed:
Updated tare default for API docs.

daversintel committed:
Merge remote-tracking branch 'origin/development' into Calib3

daversintel committed:
Merge remote-tracking branch 'origin/development' into Calib3

daversintel committed:
Merge remote-tracking branch 'origin/development' into Calib3

daversintel committed:
Merge remote-tracking branch 'origin/development' into Calib3

daversintel committed:
Merge remote-tracking branch 'origin/development' into Calib3

daversintel committed:
Don't stream 256x144 LR any more.

daversintel committed:
Version updated to 2.45.33.2.

daversintel committed:
Fixed a typo.

daversintel committed:
Restored to previous resolution to stream for OCC, OCC-exteneded, and tare.

daversintel committed:
Updated version to 2.45.33.3.

daversintel committed:
Merge remote-tracking branch 'origin/development' into Calib3

daversintel committed:
Merge remote-tracking branch 'origin/development' into Calib3

daversintel committed:
Merge remote-tracking branch 'origin/development' into Calib3

daversintel committed:
uvmapping calib eligiblility check.

daversintel committed:
1. Updated vdersion to 2.45.33.4.
2. Log preset for tare calibration.

daversintel committed:
Trying to fix python build.
Trying to fix more wrapper build errors.
Trying to fix linux build errors.

daversintel committed:
Merge remote-tracking branch 'origin/development' into Calib3

daversintel committed:
Fixed merge problem.

daversintel committed:
Added both sides option for target based focal length calibration.

daversintel committed:
UPdated version to 2.45.33.5.

daversintel committed:
Merge remote-tracking branch 'origin/development' into Calib3

daversintel committed:
Updated UCAL menu order.

daversintel committed:
Updated version to 2.45.33.6.

daversintel committed:
Added version check for tare health numbers.

daversintel committed:
Some code update based on reviewer's feedback.

daversintel committed:
Enabled tare health check numbers starting fom firmware 5.12.14.100.

daversintel committed:
Merge remote-tracking branch 'origin/development' into Calib3

# Conflicts:
#	include/librealsense2/rs.h

doronhi committed:
Fix python compilation for new calibration values.

add comments
Fix some warnings

doronhi committed:
move calculation of sum of pixels in depth image to function: get_depth_frame_sum

doronhi committed:
Merge branch 'Calib3' into calib3_doron

doronhi committed:
fix merge.

doronhi committed:
Add function add_calibration_frame for hast assisted calibration.

on-chip-calib.cpp: calibrate now uses calib_dev.add_calibration_frame for tare_calibration.
fix depth_auto_calibration_example.py to use either host_assistance tare calibration or firware-only calibration.

doronhi committed:
add_calibration_frame works with OCC.

Fix realsense-viewer and depth_auto_calibration_example.py to demonstrate usage.

doronhi committed:
update progress.

fix health.

doronhi committed:
fixed "float* health" to "float* const health" (const pointer to float)

fix return health value to python
modified the callback in depth_auto_calibration_example.py to a more graphical one.

doronhi committed:
clean

ev-mp commented on src/ds5/ds5-auto-calibration.cpp:
```diff
@@ -28,6 +28,24 @@ namespace librealsense
         float rightRotation[9]; // Right rotation
     };
 
+    struct TareCalibrationResult
+    {
+      uint16_t status;  // DscStatus
+      uint32_t tareDepth;  // Tare depth in 1/100 of depth unit
+      uint32_t aveDepth;  // Average depth in 1/100 of depth unit
+      int32_t curPx;    // Current Px in 1/1000000 of normalized unit
+      int32_t calPx;    // Calibrated Px in 1/1000000 of normalized unit
+      float curRightRotation[9]; // Current right rotation
+      float calRightRotation[9]; // Calibrated right rotation
+      uint16_t accuracyLevel;  // [0-3] (Very High/High/Medium/Low)
+      uint16_t iterations;        // Number of iterations it took to converge
+      //int32_t errors[iterations];  // Array of errors in 1/1000000 of a percent
+      //int32_t x[iterations];    // Intrinsic scan: array of Px in 1/1000000 normalized unit
+      //                         // Extrinsic scan: array of Ry in 1/100000 radian
+      //float beforeHealthCheck; // Before health check number
+      //float afterHealthCheck;  // After health check number
+    };
+
```
Verify it is handled with `#pragma pack(push, 1)`

doronhi replied:
It is in a section begins with "#pragma pack(push, 1)" and ends with "#pragma pack(pop)"

ev-mp commented on src/ds5/ds5-auto-calibration.h:
```diff
@@ -39,8 +55,30 @@ namespace librealsense
+        int _collected_frame_num;
+        double _collected_sum;
+        auto_calib_action _action;
+        interactive_calibration_state _occ_state;
```
Rename `occ` to `interactive` or similar

doronhi replied:
Done.

ev-mp commented on src/ds5/ds5-auto-calibration.cpp:
```diff
@@ -294,84 +321,166 @@ namespace librealsense
+            _occ_state = RS2_OCC_STATE_WAIT_TO_CAMERA_START;
+            switch (speed)
+            {
+            case 0:
```
There should be enum that defines the scan speed. Pls refactor with that instead of plain numbers

doronhi replied:
Done.

ev-mp commented on src/ds5/ds5-auto-calibration.cpp:
```diff
@@ -294,84 +321,166 @@ namespace librealsense
 
             // Begin auto-calibration
-            _hw_monitor->send(command{ ds::AUTO_CALIB, py_rx_calib_begin, speed, 0, param.param_4 });
+            if (host_assistance == 0 || host_assistance == 1)
```
Same here

doronhi replied:
Done.

ev-mp commented on src/ds5/ds5-auto-calibration.cpp:
```diff
@@ -294,84 +321,166 @@ namespace librealsense
+                    }
 
-            DirectSearchCalibrationResult result{};
+                    _hw_monitor->send(cmd);
```
Check the return status to handle FW-reported failures

ev-mp commented on src/ds5/ds5-auto-calibration.cpp:
```diff
@@ -294,84 +321,166 @@ namespace librealsense
-                {
-                    if (!((retries++) % 5)) // Add log debug once in a sec
+                        if (res.size() < sizeof(DirectSearchCalibrationResult))
+                            throw std::runtime_error("Not enough data from CALIB_STATUS!");
```
Replace throws with branching. We don't want to abuse the error handling for flow control

doronhi replied:
Done.

ev-mp commented on src/ds5/ds5-auto-calibration.cpp:
```diff
@@ -395,65 +504,101 @@ namespace librealsense
 
-                now = std::chrono::high_resolution_clock::now();
+                        if (res.size() < sizeof(FocalLengthCalibrationResult))
+                            throw std::runtime_error("Not enough data from CALIB_STATUS!");
```
Same as above

doronhi replied:
Done.

ev-mp reviewed (changes requested):
WIP, so far has reached ds5-auto-calibration.cpp only

doronhi committed:
use existing function to compare FW version with a given minimal one.

doronhi commented on common/on-chip-calib.cpp:
```diff
@@ -862,24 +956,31 @@ namespace rs2
         auto calib_dev = _dev.as<auto_calibrated_device>();
         if (action == RS2_CALIB_ACTION_TARE_CALIB)
-            _new_calib = calib_dev.run_tare_calibration(ground_truth, json, [&](const float progress) {_progress = progress;}, 5000);
+            _new_calib = calib_dev.run_tare_calibration(ground_truth, json, health, [&](const float progress) {_progress = progress; }, 5000);
```
This may be a new API but that's because this is a modified algorithm that now returns "health". There is no longer an API without the "health" parameter so there is no need for a different function name.

doronhi committed:
Add timeout of 30 seconds to host assistant calibration process.

doronhi commented on common/on-chip-calib.cpp:
```diff
@@ -862,24 +956,31 @@ namespace rs2
+        while (!calib_done)
+        {
+            rs2::depth_frame f = fetch_depth_frame(invoke, frame_fetch_timeout_ms);
+            _new_calib = calib_dev.add_calibration_frame(f, health, [&](const float progress) {_progress = progress; }, 5000);
```
It is caught and displayed as a pop-up notice on the top-right side of the viewer.

doronhi commented on common/ux-window.cpp:
```diff
@@ -262,7 +262,7 @@ namespace rs2
 
         rs2_error* e = nullptr;
-        _title_str = to_string() << _title << " v" << api_version_to_string(rs2_get_api_version(&e));
+        _title_str = to_string() << _title << " v" << RS2_API_FULL_VERSION_STR; // api_version_to_string(rs2_get_api_version(&e));
```
Reverted.

doronhi committed:
convert speed to enum

doronhi committed:
Merge remote-tracking branch 'origin/511' into calib3_doron

# Conflicts:
#	common/on-chip-calib.cpp
#	src/auto-calibrated-device.h
#	src/ds5/ds5-auto-calibration.h
#	src/rs.cpp

doronhi committed:
convert host_assistance to enum

validate speed value

doronhi committed:
replace try-catch sections with if-else sections as flow control handlers.

doronhi commented on common/on-chip-calib.cpp:
```diff
@@ -27,6 +28,25 @@ namespace rs2
+            }
+
+            versions.emplace_back(atoi(fw_version.substr(i).c_str()));
+            if (versions[0] >= 5 && versions[1] >= 12 && versions[2] >= 14 && versions[3] >= 100) // FW 05.12.14.100
```
Done.

doronhi commented on common/on-chip-calib.cpp:
```diff
@@ -779,6 +801,75 @@ namespace rs2
+                if (*p)
+                {
+                    ++count;
+                    sum += *p;
```
Max_uint16 = 65535
According to [This question](https://stackoverflow.com/questions/1848700/biggest-integer-that-can-be-stored-in-a-double): Max_double = 9007199254740991
Max_sum_of_image = 65535*1280*800
Max_double / Max_sum_of_image= 134219.77603125045

So, as long as we sum less than 134000 images (The current algorithm sums up 20 images), "sum" cannot overflow.

doronhi commented on common/on-chip-calib.cpp:
```diff
@@ -862,24 +956,31 @@ namespace rs2
+        int frame_fetch_timeout_ms = 3000;
+
+        bool calib_done(!_new_calib.empty());
+        while (!calib_done)
```
Done

doronhi commented on common/on-chip-calib.cpp:
```diff
@@ -1243,6 +1355,27 @@ namespace rs2
+            0x00, 0x00, 0x00, 0x00,
+            0x00, 0x00, 0x00, 0x00
+        };
+
```
Merged with branch 511. Function is removed.

doronhi committed:
Add comment for _health1, _health2 parameters.

doronhi commented on common/on-chip-calib.cpp:
```diff
@@ -895,6 +996,15 @@ namespace rs2
+        else if (action == RS2_CALIB_ACTION_TARE_CALIB)
+        {
+            _health_1 = health[0] * 100;
+            _health_2 = health[1] * 100;
```
Done.

doronhi committed:
fix bug: Crash if asked firmware for status while firmware was in progress.

doronhi committed:
ask firmware for valid depth values.

doronhi committed:
Throw if firmware values are not yet set.

doronhi committed:
Fix ROI scan.

doronhi committed:
DEBUG: Save raw images

doronhi committed:
add "resize factor" element to input json string - int. cropped size is image_size / resize_factor. i.e. 5 - means work on central width/5, height/5 image of the image.

doronhi committed:
fix bug in ROI calculation in host assistance tare calibration in collect_depth_frame_sum

Save images for host assistance tare calibration
Save all images with indexes

doronhi committed:
fix condition of pixel validity to include threshold values.

doronhi committed:
Fix passing exception messages to python wrapper.

Fix depth_auto_calibration_example.py import for linux.

doronhi committed:
Merge remote-tracking branch 'doronhi/calib3_doron' into calib3_doron

doronhi committed:
remove redundant parameters in depth_auto_calibration_example.py

remove log messages.

doronhi committed:
Save fill_factor_final.txt

doronhi committed:
Merge branch 'calib3_doron' of https://github.com/doronhi/librealsense into calib3_doron

# Conflicts:
#	wrappers/python/examples/depth_auto_calibration_example.py

doronhi committed:
shift values in _fill_factor table by 1 - It seems that for frame_counter=1 we get the image of frame_counter 0 etc. This fix is meant to compensate for that.

doronhi committed:
skip first frame in tare calculations.

doronhi committed:
Enable option for "interactive scan" - user should set "interactive scan" to 1 to enable.

Add debug file to working directory: all_frame_numbers.txt with the following format: frame_number, prev_frame_number, total_frame, frame_time, computer_time
Add timestamps to saved debug images filenames - to avoid overriding.

doronhi committed:
For OCC: Add offset of +1 when using host with iterative mode and +2 with host without iterative mode. (fill_rate[idx+offset] = calc_fill_rate(idx) )

doronhi committed:
Reduce Offset to 0 in interactive mode and 1 in regular mode.

doronhi committed:
Add interactive key (i)

doronhi committed:
Remove saving debug images and files

realsense-viewer default mode for host-assisted OCC changed to interactive.
rename rs2_add_calibration_frame to rs2_process_calibration_frame

doronhi committed:
fix converged typo

Nir-Az commented:
@ev-mp Is this still relevant?

Nir-Az commented:
Thank you for this contribution, and our apologies for the long delay in getting back to you.

This PR has been awaiting a response from the author for an extended period (over a year in most cases), and the codebase has moved on significantly since it was opened. To keep the open pull-request list focused on actionable work, we're closing it for now.

This isn't a reflection on the quality of the contribution. If it's still relevant and you'd like to continue, please feel free to reopen it or open a fresh PR rebased on the latest codebase — we'd be glad to take another look. Thanks again for your effort and understanding.

🤖 Generated by AI
